### DISCUSSION TITLE
Why are modelines forceably disabled?

### DISCUSSION THREAD

**Discussion by adamjhawley** [vscode-neovim/vscode-neovim#1530] [General]
Hi all, thank you for all your work on such a useful project!

I work on several projects which use vim/nvim modelines to dictate basic styling. I noticed that 'modelines' are forcibly disabled as of this commit: https://github.com/vscode-neovim/vscode-neovim/commit/abcd6a97e041b93986f89c00d2dfa2d547270c90.

I read the comment in 'vim/vscode-options.vim' but didn't really understand the reasoning and am left with a couple of questions:

1) Why are modelines forcibly disabled? If a file requires different tab settings, I want them to be used over any other setting.
2) If they need to be disabled, do you have an alternative solution I can use without manually configuring attributes such as 'textwidth' for different file types?

**Comment by theol0403** [comment]
I believe this is because setting tab settings from within neovim can break the synchronization between vim and vscode, because they need to be equivalent. It may be that in reality its fine, I'm not sure. 

You can override the setting by doing:
```lua
api.nvim_create_autocmd({ "BufEnter", "FileType" }, {
  callback = function()
    vim.opt.modeline = true
    vim.opt.modelines = 5
  end,
})
```

Alternately, I think VSCode should be able to handle everything by default, through autodetection? It confgures neovim with the proper tab settings on buffer init.

**Reply by adamjhawley** [reply]
Thanks a lot for your quick response @theol0403!

> I believe this is because setting tab settings from within neovim can break the synchronization between vim and vscode, because they need to be equivalent. It may be that in reality its fine, I'm not sure.

Ok, I can see how that might be problematic. Is it worth me spending time testing 'vscode-neovim' with the forced modeline setting removed?
> You can override the setting by doing:
> 
> ```lua
> api.nvim_create_autocmd({ "BufEnter", "FileType" }, {
>   callback = function()
>     vim.opt.modeline = true
>     vim.opt.modelines = 5
>   end,
> })
> ```

Where can I use this? Should I modify my own installation of the 'vscode-neovim' extension or can I somehow add the autocmd in a setting/config file?
> 
> Alternately, I think VSCode should be able to handle everything by default, through autodetection? It confgures neovim with the proper tab settings on buffer init.

Looks like VSCode will autodetect the language and take format settings from there. But I have multiple projects with different styling requirements for the same language, so this won't solve it for me unfortunately.

**Reply by theol0403** [reply]
> Is it worth me spending time testing 'vscode-neovim' with the forced modeline setting removed? 

Go for it, I'm not sure if recent updates have removed the possible bugs.

> Where can I use this?

This would be on the nvim side, through init.lua (the example code) or init.vim (would need to adjust the code accordingly). There needs to be no modifications of this extension. 

> But I have multiple projects with different styling requirements for the same language, so this won't solve it for me unfortunately.

You can configure project-specific settings in vscode. Even better, you can use editorconfig files (which can also be picked up by terminal nvim).

**Comment by xiyaowong** [comment]
You can consider creating an issue so that others can support `modeline` when they have time.

**Comment by xiyaowong** [comment]
`modeline` works now. See #1543

**Reply by adamjhawley** [reply]
Thanks a lot, @xiyaowong, you created a fix before I even got round to creating the issue!

**Comment by adamjhawley** [comment]
Thanks again @theol0403 and @xiyaowong for such quick and helpful responses to my question above. I look forward to using the fix shortly!

Closing this discussion as it should now be resolved.
